core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -278,7 +278,7 @@ class Partition(val topicPartition: TopicPartition,
             if (futureLogDir != logDir)
               throw new IllegalStateException(s"The future log dir $futureLogDir of $topicPartition is " +
                 s"different from the requested log dir $logDir")
-            false
+            true
```

Comment:
@junrao thanks for the feedback!

> Do you know why the ReplicaAlterLogDirsThread died? If it's a bug there, we probably should fix the logic in ReplicaAlterLogDirsThread.

1. send request to alter replica dir
1. future log is created
1. ReplicaAlterLogDirsThread is created and it gets epoch N
1. send request to change partition reassignment (but the new reassignment is same to current reassignment)
1. KafkaController assumes the reassignment is completed
1. LeaderAndIsrRequest is sent
1. ReplicaManager#becomeLeaderOrFollower updates epoch of partition (from N to N + 1)
1. ReplicaAlterLogDirsThread encounters FENCED_LEADER_EPOCH since its fetch request is stale
1. ReplicaAlterLogDirsThread move the partition to failedPartitions
1. ReplicaAlterLogDirsThread get idle
1. ReplicaManager#becomeLeaderOrFollower shutdowns idle ReplicaAlterLogDirsThread

Since the Partition.futureLog is existent, RepliaManager#alterReplicaLogDirs does NOT add ReplicaAlterLogDirsThread for requests.

```scala
          if (partition.maybeCreateFutureReplica(destinationDir, highWatermarkCheckpoints)) {
            val futureLog = futureLocalLogOrException(topicPartition)
            logManager.abortAndPauseCleaning(topicPartition)

            val initialFetchState = InitialFetchState(BrokerEndPoint(config.brokerId, "localhost", -1),
              partition.getLeaderEpoch, futureLog.highWatermark)
            replicaAlterLogDirsManager.addFetcherForPartitions(Map(topicPartition -> initialFetchState))
          }
```

Reply:
@chia7712 : Thanks for the reply. Great finding! This seems to be a bug. So, in step 7, whenever we bump up the leader epoch in ReplicaManager#becomeLeaderOrFollower(). It seems that we should update the leader epoch for existing partitions in ReplicaAlterLogDirsThread. One way to do that is to first call replicaAlterLogDirsManager.removeFetcherForPartitions() and then addFetcherForPartitions() on that partition.

Reply:
@junrao thanks for the suggestions!

This issue is still existent even if we updates the partitions in ReplicaManager#becomeLeaderOrFollower. For example, the following execution order causes the updated partition is removed.
1. ReplicaManager#updates the leader epoch of partition
1. ReplicaAlterLogDirsThread gets epoch error due to (local) stale leader epoch
1. ReplicaManager#updates the leader epoch of ReplicaAlterLogDirsThread
1. ReplicaAlterLogDirsThread remove fenced partition and then get idle

It seems to me the root cause is that we do not change both parittion epoch and ReplicaAlterLogDirsThread epoch atomically. Hence, removing the fenced partition from ReplicaAlterLogDirsThread is dangerous to this issue. If we want to make ReplicaAlterLogDirsThread be recoverable from fenced error, the fenced partition should be added back as "delay" partition. Also, ReplicaManager#becomeLeaderOrFollower should update the epoch of ReplicaAlterLogDirsThread.

Reply:
@chia7712 : Thanks for the reply. If a partition is removed (or re-inserted) from AbstractFetcherThread (which ReplicaAlterLogDirsThread inherits from), processFetchRequest() will ignore any pending fetch response for that partition. So, the scenario that you described should be handled even though we don't have the logic to change epoch atomically in multiple places.

Reply:
> If a partition is removed (or re-inserted) from AbstractFetcherThread (which ReplicaAlterLogDirsThread inherits from), processFetchRequest() will ignore any pending fetch response for that partition

Thanks for the feedback! This comment is totally right but the key point related to this issue is the partition is removed by AbstractFetcherThread#processFetchRequest().

```scala
                case Errors.FENCED_LEADER_EPOCH =>
                  onPartitionFenced(topicPartition)
```

After the partition is removed, the thread becomes idle and then ReplicaManager#becomeLeaderOrFollower remove it.

```scala
        replicaAlterLogDirsManager.addFetcherForPartitions(futureReplicasAndInitialOffset)

        replicaFetcherManager.shutdownIdleFetcherThreads()
        replicaAlterLogDirsManager.shutdownIdleFetcherThreads()  // this line
```

And noted that ```futureReplicasAndInitialOffset``` is NOT empty only if there is "new" partition. In this case, there is no new partition so nothing is re-added to ```replicaAlterLogDirsManager```.

In short, the thread is shutdown and it leaves a future folder.

Another issue is ReplicaManager#alterReplicaLogDirs add new partition to ReplicaAlterLogDirsManager only if the future folder is NOT existent (and it is created successfully)

```scala
          if (partition.maybeCreateFutureReplica(destinationDir, highWatermarkCheckpoints)) {
            ...
            replicaAlterLogDirsManager.addFetcherForPartitions(Map(topicPartition -> initialFetchState))
          }
```

So the following request to alter replica folder don't work util the future replica is removed.